x/bep3/types/msg.go

```diff
@@ -0,0 +1,254 @@
+func (msg MsgCreateAtomicSwap) Route() string { return RouterKey }
+
+// Type is the name of MsgCreateAtomicSwap
+func (msg MsgCreateAtomicSwap) Type() string { return CreateAtomicSwap }
+
+// String prints the MsgCreateAtomicSwap
+func (msg MsgCreateAtomicSwap) String() string {
+	return fmt.Sprintf("AtomicSwap{%v#%v#%v#%v#%v#%v#%v#%v#%v#%v}",
+		msg.From, msg.To, msg.RecipientOtherChain, msg.SenderOtherChain,
+		msg.RandomNumberHash, msg.Timestamp, msg.Amount, msg.ExpectedIncome,
+		msg.HeightSpan, msg.CrossChain)
+}
+
+// GetInvolvedAddresses gets the addresses involved in a MsgCreateAtomicSwap
+func (msg MsgCreateAtomicSwap) GetInvolvedAddresses() []sdk.AccAddress {
+	return append(msg.GetSigners(), AtomicSwapCoinsAccAddr)
+}
+
+// GetSigners gets the signers of a MsgCreateAtomicSwap
+func (msg MsgCreateAtomicSwap) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{msg.From}
+}
+
+// ValidateBasic validates the MsgCreateAtomicSwap
+func (msg MsgCreateAtomicSwap) ValidateBasic() sdk.Error {
+	if len(msg.From) != AddrByteCount {
+		return sdk.ErrInternal(fmt.Sprintf("the expected address length is %d, actual length is %d", AddrByteCount, len(msg.From)))
+	}
+	if len(msg.To) != AddrByteCount {
+		return sdk.ErrInternal(fmt.Sprintf("the expected address length is %d, actual length is %d", AddrByteCount, len(msg.To)))
+	}
+	if !msg.CrossChain && len(msg.RecipientOtherChain) != 0 {
+		return sdk.ErrInternal(fmt.Sprintf("must leave recipient address on other chain to empty for single chain swap"))
+	}
+	if !msg.CrossChain && len(msg.SenderOtherChain) != 0 {
+		return sdk.ErrInternal(fmt.Sprintf("must leave sender address on other chain to empty for single chain swap"))
+	}
+	if msg.CrossChain && len(msg.RecipientOtherChain) == 0 {
+		return sdk.ErrInternal(fmt.Sprintf("missing recipient address on other chain for cross chain swap"))
+	}
+	if len(msg.RecipientOtherChain) > MaxOtherChainAddrLength {
+		return sdk.ErrInternal(fmt.Sprintf("the length of recipient address on other chain should be less than %d", MaxOtherChainAddrLength))
+	}
+	if len(msg.SenderOtherChain) > MaxOtherChainAddrLength {
+		return sdk.ErrInternal(fmt.Sprintf("the length of sender address on other chain should be less than %d", MaxOtherChainAddrLength))
+	}
+	if len(msg.RandomNumberHash) != RandomNumberHashLength {
+		return sdk.ErrInternal(fmt.Sprintf("the length of random number hash should be %d", RandomNumberHashLength))
+	}
+	if !msg.Amount.IsAllPositive() {
+		return sdk.ErrInternal(fmt.Sprintf("the swapped out coin must be positive"))
+	}
+	if len(msg.ExpectedIncome) > MaxExpectedIncomeLength {
+		return sdk.ErrInternal(fmt.Sprintf("the length of expected income should be less than %d", MaxExpectedIncomeLength))
+	}
+	expectedIncomeCoins, err := sdk.ParseCoins(msg.ExpectedIncome)
+	if err != nil || expectedIncomeCoins == nil {
+		return sdk.ErrInternal(fmt.Sprintf("expected income %s must be in valid format e.g. kava10000", msg.ExpectedIncome))
```

Comment:
```suggestion
		return sdk.ErrInternal(fmt.Sprintf("expected income %s must be in valid format e.g. 10000ukava", msg.ExpectedIncome))
```

nit: match cosmos sdk parsing rather than bnbchain

Reply:
Fixed in https://github.com/Kava-Labs/kava/pull/370/commits/ac92e26999cf32a5fae8296ba92a05b1da0cf357